### DISCUSSION TITLE
IEnumerable<KeyValuePair<string, string?>> and IEnumerable<KeyValuePair<string, string>>

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4894] [Q&A]
How do I write a method that can accept an:
```IEnumerable<KeyValuePair<string, string?>>```
or an
```IEnumerable<KeyValuePair<string, string>>```
as a parameter value?

Because string is a class, they boil down to the same thing except I can't seem to make the nullable warnings go away.

**Comment by huoyaoyuan** [comment]
Currently there's no way. It's similar to the problem of `Task<T?>`.

**Reply by jnm2** [reply]
I think the problem stems from structs and classes not having generic variance. You can't assign a `KeyValuePair<string, string>` to `KeyValuePair<object, object>` or a `Task<string>` to a `Task<object>`, and the same invariance holds for `KeyValuePair<string, string>` -> `KeyValuePair<string?, string?>`.

**Reply by 333fred** [reply]
That's correct.

**Reply by TonyValenti** [reply]
It seems odd to be though because a KeyValuePair<string, string> is the same CLR type as a KeyValuePair<string?, string?>. It seems like in a situation like that, a warning shouldn't exist. Especially since a keyvaluepair is  essentially all out parameters except for the constructor.

**Reply by jnm2** [reply]
The trouble is that there's no way to annotate whether TKey and TValue are `out` or `in` parameters, and the compiler can't just assume one or the other.

**Comment by RikkiGibson** [comment]
"Task<T> nullability covariance" #3950 is approved but needs design. I feel that the feature should be extended into a more generalized "nullable variance" feature. KeyValuePair is a particular pain point, and so are tuples and other kinds of generic containers.

**Comment by ufcpp** [comment]
Real world issue:

```cs
using System.Collections.Generic;
using System.Net.Http;

#nullable enable
_ = new FormUrlEncodedContent(new Dictionary<string, string>()); // CS8620 in .NET 5
```
